fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/rule/EnumRule.java

```diff
@@ -0,0 +1,161 @@
+import java.util.Collections;
+import java.util.Iterator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.GameRules;
+
+import net.fabricmc.fabric.api.gamerule.v1.GameRuleRegistry;
+
+public class EnumRule<E extends Enum<E>> extends LiteralRule<EnumRule<E>> {
+	private static final Logger LOGGER = LogManager.getLogger(GameRuleRegistry.class);
+
+	private final Class<E> classType;
+	protected final Collection<E> supportedValues;
+	protected E value;
+
+	/**
+	 * You should not be calling this constructor!
+	 */
+	@Deprecated
+	public EnumRule(GameRules.Type<EnumRule<E>> type, E value, E[] supportedValues) {
+		this(type, value, Arrays.asList(supportedValues));
+	}
+
+	/**
+	 * You should not be calling this constructor!
+	 */
+	@Deprecated
+	public EnumRule(GameRules.Type<EnumRule<E>> type, E value, Collection<E> supportedValues) {
+		super(type);
+		this.classType = value.getDeclaringClass();
+		this.value = value;
+		this.supportedValues = Collections.unmodifiableCollection(supportedValues);
+	}
+
+	@Override
+	protected void deserialize(String value) {
+		try {
+			final E deserialized = Enum.valueOf(this.classType, value);
+
+			if (!this.supports(deserialized)) {
+				LOGGER.warn("Failed to parse rule of value {} for rule of type {}. Since the value {}, is unsupported.", value, this.classType, value);
+			}
+
+			this.set(deserialized, null);
+		} catch (IllegalArgumentException e) {
+			LOGGER.warn("Failed to parse rule of value {} for rule of type {}", value, this.classType);
+		}
+	}
+
+	@Override
+	public String serialize() {
+		return this.value.name();
+	}
+
+	@Override
+	public int getCommandResult() {
+		// For now we are gonna use the ordinal as the command result. Could be changed or set to relate to something else entirely.
+		return this.value.ordinal();
+	}
+
+	@Override
+	protected EnumRule<E> getThis() {
+		return this;
+	}
+
+	public Class<E> getEnumClass() {
+		return this.classType;
+	}
+
+	@Override
+	public String toString() {
+		return this.value.toString();
+	}
+
+	@Override
+	protected EnumRule<E> copy() {
+		return new EnumRule<E>(this.type, this.value, this.supportedValues);
+	}
+
+	@Override
+	public void setValue(EnumRule<E> rule, MinecraftServer minecraftServer) {
+		this.value = rule.value;
+		this.changed(minecraftServer);
+	}
+
+	public E get() {
+		return this.value;
+	}
+
+	public E cycle(E start) {
+		if (this.supportedValues.size() > 1) {
+			E value = getNext(this.supportedValues, start);
+
+			return value;
+		}
+
+		return start;
+	}
+
+	protected static <T> T getNext(Collection<T> values, T value) {
+		Iterator<T> iterator = values.iterator();
+
+		do {
+			if (!iterator.hasNext()) {
+				return iterator.next();
+			}
+		} while (!iterator.next().equals(value));
+
+		if (iterator.hasNext()) {
+			return iterator.next();
+		} else {
+			return values.iterator().next();
+		}
+	}
```

Comment:
Change this to a single line return with a ternary operator?
```java
return iterator.hasNext() ? iterator.next() : values.iterator().next();
```

Reply:
At least there shouldn't be an `else` statement after a `return` statement

Reply:
Yea, I've collapsed it to the ternary operator